Add Advanced HUP and update existing HUP docs (#3115)
balena-io/docs: pages/reference/OS/updates/update-process.md

1. Partition layout changes from one OS version to another. For example:
    
    -  The partition layout for NVIDIA JetPack 4-based {{ $names.os.lower }} differs from JetPack 5-based {{ $names.os.lower }}. If a device is running already running JetPack-4 {{ $names.os.lower }}, you can use advanced HUP to update to a JetPack 5-based {{ $names.os.lower }}.
    - {{ $names.company.upper }} releases a new version of an OS for a certain device type whereby the root partition has been increased to accommodate more kernel module and drivers.
    -  {{ $names.company.upper }} releases a new version of an OS for a certain device type to include new functionality such as Secure Boot.
2. Changing OS architecture, for example: 
    - Switching from a the 32-bit OS variant running on Raspberry Pi 3 to a 64-bit variant of {{ $names.os.lower }}.

__Note:__ If you've used advanced HUP, it does not necessarily mean that future updates won't be compatible with standard HUP. 

### Caveats
When using the advanced HUP, it’s important to be aware of several limitations compared to the standard HUP mechanism.

> - Data loss

Reviewer: I don't see any special formatting for these bullet points.

![image](https://github.com/user-attachments/assets/f483b889-cafa-43f3-88bd-b716876785a3)

excerpt: Details of the {{ $names.os.lower }} self-service update process
---

# {{ $names.company.upper }} host OS updates

> With OTA (Over-the-Air) functionality built in, {{ $names.company.lower }} enables you to remotely update the entire host operating system (OS) on your device, rather than just the user application. {{ $names.company.upper }} provides two methods to perform Host OS Updates (HUP):

Reviewer: Make it clear that the user does not choose which type of update to run. Instead it's the nature of the update.

Author: thanks, added some notes regarding this

1. Standard HUP: The main mechanism for updating the hostOS, providing features such as automatic rollback and delta updates.

> 2. Advanced HUP: There are certain upgrade paths where Standard HUP is not compatible. This update mechanism is usually more involved and does not provide features such as automatic rollback and delta updates. Advanced HUP is usually used for certain specific scenarios and subsequent updates can then be done via Standard HUP. For e.g. The partition layout for NVIDIA JetPack 4-based {{ $names.company.lower }}OS differs from JetPack 5-based {{ $names.company.lower }}OS. If a device is running already running JetPack-4 {{ $names.company.lower }}OS, you can use Advanced HUP to update to a JetPack 5-based {{ $names.company.lower }}OS. Subsequent updates can be carried out using Standard HUP.

Reviewer: Text can be tightened up, and "For e.g." is not correct. I would just say "For example".

excerpt: Details of the {{ $names.os.lower }} self-service update process
---

# {{ $names.company.upper }} host OS updates

With OTA (Over-the-Air) functionality built in, {{ $names.company.lower }} enables you to remotely update the entire host operating system (OS) on your device, rather than just the user application. {{ $names.company.upper }} provides two methods to perform Host OS Updates (HUP):

1. Standard HUP: The main mechanism for updating the hostOS, providing features such as automatic rollback and delta updates.

> 2. Advanced HUP: There are certain upgrade paths where Standard HUP is not compatible. This update mechanism is usually more involved and does not provide features such as automatic rollback and delta updates. Advanced HUP is usually used for certain specific scenarios and subsequent updates can then be done via Standard HUP. For example, the partition layout for NVIDIA JetPack 4-based {{ $names.os.lower }} differs from JetPack 5-based {{ $names.os.lower }}. If a device is running already running JetPack-4 {{ $names.os.lower }}, you can use Advanced HUP to update to a JetPack 5-based {{ $names.os.lower }}. Subsequent updates can be carried out using Standard HUP.

Reviewer 2: ```suggestion
2. Advanced HUP: Standard HUP can be incompatible for certain upgrade paths. This update mechanism is usually more involved and does not provide features such as automatic rollback and delta updates. Advanced HUP is used for specific scenarios. Subsequent updates to the OS can be done through standard HUP. For example, the partition layout for NVIDIA JetPack 4-based {{ $names.os.lower }} differs from JetPack 5-based {{ $names.os.lower }}. If a device is running already running JetPack-4 {{ $names.os.lower }}, you can use Advanced HUP to update to a JetPack 5-based {{ $names.os.lower }}. After the advanced HUP, the next updates can be carried out using the standard HUP process.
```

> You cannot select the mechanism to be used for the update. Rather, {{ $names.company.lower }} determines the mechanism to be used based on the source OS version currently running on the device and the target OS version selected.

Reviewer 2: ```suggestion
The method used to perform the host OS update doesn't need to be selected. Rather, {{ $names.company.lower }} determines the method to be used automatically. This is based on the current OS version running on the device and the target OS version selected.
```

## Standard HUP

> 

Reviewer 2: Might we add one line to intro this method first?

Reviewer 2: Reason: If this heading gets linked from somewhere else, it should make sense atomically too without needing more context

Author: i did, but it's getting quite repetitive

### Features

#### Atomic Updates
In general, host OS updates are meant to be atomic: if an update is run, it either finishes correctly or it fails and leaves the system in its previous, usable state. If deploying the new root file system is unsuccessful, the boot settings are not switched over to the new partition. This means the device can be rebooted with the previous OS version and no noticeable changes.

__Note:__ Once a host OS update is successful, it is not possible to roll back to a previous OS version (except via [automatic rollbacks](#automatic-rollbacks) as noted below).

#### Automatic Rollback
For failures related to the boot partition, the latest versions of {{ $names.os.lower }} have a rollback feature that will leave the partition in a good state.

#### Delta Updates and Bandwidth Efficiency
A delta update contains only the differences between the current OS and the new version. This reduces the amount of data that needs to be transferred and can significantly lower network costs and update times. 

> ### How it works

Reviewer 2: move this section above features, it seems like we are describing how Delta updates work by keeping it below.

> In the update process, an updater script gets transferred to the device to handle the update process, as described in detail below.

Reviewer 2: redundant.

Reviewer 2: see below

> The first step the updater script performs includes a number of coherence checks, so that an update does not progress if something isn't right. The checks include, among other things, whether the device is running a {{ $names.os.lower }} version that it can work with, whether the script is running on the right device type, and whether it can reach the {{ $names.cloud.lower }} registry to get the target {{ $names.os.lower }} image.

Reviewer 2: ```suggestion
After a host OS update (HUP) is triggered, an updater script gets transferred to the device from the backend to handle the update process. The updater script performs a number of coherence checks, so an update does not progress if something isn't right. The checks include, among other things, whether the device is running a {{ $names.os.lower }} version that it can work with, whether the script is running on the right device type, and whether it can reach the {{ $names.cloud.lower }} registry to get the target {{ $names.os.lower }} image.
```

The updater script checks to see if the new {{ $names.os.lower }} version ships with a newer supervisor version than what the device currently runs, and, if so, the supervisor is updated.

Finally, the boot settings are modified so that on the next reboot the new root file system is used. As a last step, the device is rebooted.

### Logging
A status of `OS update failed` means the user application should still be running normally, and the reasons for failure can be examined through the update logs at `/mnt/data/balenahup/` (or potentially the legacy location `/mnt/data/resinhup/`). The device may have some reduced functionality, for instance if the supervisor was stopped for the update, but we are working on ways to automatically restore full functionality whenever possible.

> ### Automatic rollbacks

Reviewer 2: Why 2 different sections to explain rollback. 
Move the "How it works section" above features
Merge the feature rollback with this rollback

Rollbacks is a framework designed to automatically roll back the OS update in case something goes wrong.

There are two rollback mechanisms in the OS, covering different update failure modes. One based on health checks called the `rollback-health`, and the recognizing if the new system is unbootable for some reason, the `rollback-altboot`. Learn more about [rollback mechanisms][rollback-mechanism].

## Advanced HUP

> As mentioned above, Advanced HUP is usually required for certain very specific upgrade paths. The underlying reasons for incompatibility with Standard HUP can be, but not limited to:

Reviewer 2: ```suggestion
Advanced HUP is required for only certain upgrade paths. The underlying reasons for incompatibility with standard HUP can be, but not limited to:
```

1. Partition layout changes from one OS version to another. For example:
    
    -  The partition layout for NVIDIA JetPack 4-based {{ $names.os.lower }} differs from JetPack 5-based {{ $names.os.lower }}. If a device is running already running JetPack-4 {{ $names.os.lower }}, you can use Advanced HUP to update to a JetPack 5-based {{ $names.os.lower }}
    - {{ $names.company.upper }} releases a new version of an OS for a certain device type whereby the root partition has been increased to accommodate more kernel module and drivers

>     -  {{ $names.company.upper }} releases a new version of an OS for a certain device type to include new functionality such as Secure Boot

Reviewer 2: Please add full stops after each each point

> 2. Changing OS architecture, for example: 
>     - Switching from a the 32-bit OS variant running on Raspberry Pi 3 to a 64-bit variant of {{ $names.os.lower }}
> 
> __Note:__ If you've used Advanced HUP, it does not mean necessarily updates won't be compatible with Standard HUP. For instance, if a device running 32-bit  {{ $names.os.lower }} on a Raspberry Pi 3 has been updated to a 64-bit variant of {{ $names.os.lower }}, Standard HUP can be used for future updates to the 64-bit OS. In this case, Advanced HUP is only used once to migrate the device from one OS architecture to another.

Reviewer 2: ```suggestion
2. Changing OS architecture, for example: 
    - Switching from a 32-bit OS variant running to a 64-bit variant of {{ $names.os.lower }}. Standard HUP can be used for future updates to the 64-bit OS. In this case, Advanced HUP is only used once to migrate the device from one OS architecture to another.
```

### Caveats

> When using the Advanced HUP, it’s important to be aware of several limitations compared to the Standard HUP mechanism.

Reviewer 2: capitals HUP methods needs to be fixed

> #### Data loss

Reviewer 2: These can be points, no need for headings

Data loss may occur, as the process can involve overwriting partitions and filesystems. Please make sure to backup all data on the device before running any update. Also, after a successful update, all application containers will be downloaded again.

#### No automatic rollbacks
Advanced HUP does not provide a built-in rollback mechanism. If the new OS version proves unstable or introduces unexpected issues, there is no automated way to revert to a previous state. 

#### No delta updates
The OS image is downloaded directly via the ap. The registry and {{ $names.engine.lower }} is not used during the download process.

### How it works

> Similar to Standard HUP, an updater script gets transferred to the device to handle the update process.

Reviewer 2: The docs are better kept atomic, referring to a previous section might not be required here. Since we are fully describing the process. 

If you accepted the previous suggestion, then you can make the same improvement here.

The first step the updater script performs includes a number of coherence checks, so that an update does not progress if something isn't right. The checks include, among other things, whether the device is running a {{ $names.os.lower }} version that it can work with, whether the script is running on the right device type, and whether it can reach the {{ $names.cloud.lower }} api to get the target {{ $names.os.lower }} image.

> Next, the updater script downloads the required OS image from the api. This differs from Standard HUP whereby the OS image is downloaded from the registry and is the reason why Advanced HUP does not provide delta updates. The updater script then downloads the [takeover binary][takeover] which is responsible for the main update process.

Reviewer: I think "takeover" needs some introduction:

The updater script then downloads the Advanced update executable, known as ["takeover"][takeover], which performs the update.

The updater script hands over to `takeover` which:

> - kills running processes
> - copies required files and artifacts to RAM

Reviewer 2: ```suggestion
- Kills running processes
- Copies required files and artifacts to RAM
```

- Unmounts partitions
- Flashes the target OS image to disk already running {{ $names.os.lower }}
- Transfers required files that were copied to RAM to their respective destinations on the newly flashed OS

The device is then rebooted.

> You can read more about the tool [here](https://github.com/balena-os/takeover?tab=readme-ov-file#how-it-works).

Reviewer 2: This way of linking the docs is non-accessible. Since readers who are visually challenged can't figure out the visual semantics of "here"

https://ux.stackexchange.com/a/12130

```suggestion
Learn more about [Takeover][takeover] tool and how it works.
```

> __Note:__ Because the disk running {{ $names.os.lower }} is overwritten with the target OS image, there is no automated rollback mechanism when using Advanced HUP. However, if the process is interrupted before the disk is overwritten, the device is rebooted and it will boot into the OS that was already running.

Reviewer 2: Not ideal to start a note and a new sentence with `because`

```suggestion
Since the disk running {{ $names.os.lower }} is overwritten with the target OS image during advanced HUP, there is no automated rollback mechanism. However, if the process was interrupted before the disk was overwritten, the device reboots into the OS that was already running.
```

excerpt: Details of the {{ $names.os.lower }} self-service update process
---

# {{ $names.company.upper }} host OS updates

With OTA updates (Over-the-Air) functionality built-in, {{ $names.os.lower }} enables you to remotely update the host operating system on your device, apart from the user application. {{ $names.os.upper }} provides two methods to perform host OS Updates (also, called HUP):

1. Standard HUP: The main mechanism for updating the hostOS, providing features such as [automatic rollbacks](#automatic-rollbacks) and [delta updates](#delta-updates-and-bandwidth-efficiency).

> 2. Advanced HUP: Standard HUP can be incompatible for certain upgrade paths. This update mechanism is usually more involved and does not provide features such as automatic rollback and delta updates. Advanced HUP is used for specific scenarios. Subsequent updates to the OS can be done through standard HUP. For example, the partition layout for NVIDIA JetPack 4-based {{ $names.os.lower }} differs from JetPack 5-based {{ $names.os.lower }}. If a device is already running JetPack-4 {{ $names.os.lower }}, you can use advanced HUP to update to a JetPack 5-based {{ $names.os.lower }}. After the advanced HUP, the next updates can be carried out using the standard HUP process.

Reviewer 2: ```suggestion
2. Advanced HUP: Standard HUP can be incompatible with certain upgrade paths. For these specific scenarios, we have the advanced HUP method to update devices. An example scenario would be updating from NVIDIA JetPack 4-based {{ $names.os.lower }}  to JetPack 5-based {{ $names.os.lower }}. This scenario would need advanced HUP for the update method due to a difference in partition layout. After the advanced HUP, the next updates can be carried out using the standard HUP process.
```

Suggestion: Shortenned the definition for a better understanding flow and less repetition of the same info.

The updater script checks to see if the new {{ $names.os.lower }} version ships with a newer supervisor version than what the device currently runs, and, if so, the supervisor is updated.

Finally, the boot settings are modified so that on the next reboot the new root file system is used. As a last step, the device is rebooted.

### Features

#### Atomic Updates
In general, host OS updates are meant to be atomic: if an update is run, it either finishes correctly or it fails and leaves the system in its previous, usable state. If deploying the new root file system is unsuccessful, the boot settings are not switched over to the new partition. This means the device can be rebooted with the previous OS version and no noticeable changes.

> __Note:__ Once a host OS update is successful, it is not possible to roll back to a previous OS version (except via [automatic rollbacks](#automatic-rollbacks) as noted below).

Reviewer 2: ```suggestion
__Note:__ Once a host OS update is successful, it is not possible to roll back to a previous OS version.
```

Automatic rollbacks shouldn't be suggested as any way to roll back the OS. It can be an inaccurate precedent because iirc you can't trigger an automatic rollback manually to revert the OS update.
